PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc

```diff
@@ -1097,27 +1097,24 @@ void PATMuonProducer::embedHighLevel(pat::Muon& aMuon,
   // Correct to PV
 
   // PV2D
+  aMuon.setDB(track->dxy(primaryVertex.position()),
+              track->dxyError(primaryVertex.position(), primaryVertex.covariance()),
+              pat::Muon::PV2D);
+
+  // PV3D
   std::pair<bool, Measurement1D> result =
-      IPTools::signedTransverseImpactParameter(tt, GlobalVector(track->px(), track->py(), track->pz()), primaryVertex);
+      IPTools::signedImpactParameter3D(tt, GlobalVector(track->px(), track->py(), track->pz()), primaryVertex);
```

Comment:
doesn't the 3D also have the same problem as the 2D case?
Also, does this agree (in a linear approximation) with the reco::Track::dsz ?

Reply:
Yes, the difference in the sign between that of the best track and the one obtained by IPTools remains. Nonetheless, the sign itself is not used elsewhere as that of the 2D version is. I understand that using track information would be preferred from the usage of computational resources point of view?

From a quick set of tests I see, surprisingly, that in most cases (~90%) both computations of the 3D IP agree, but there is a significant set of deviations between the tracks' dsz and iptools' value for ~10% of the events. This seems to be correlated with the linear approximation but not completely caused by it (i.e. it happens for several cases in which vertex ~ (0,0,0) and with relatively high pT/low bending). 

The sip3D = abs(d3d)/error(d3d) variable is used down the line in terms of ID criteria. I.e. for the prompt MVA discriminant (https://github.com/cms-sw/cmssw/blob/e4c6faa3322fcebf71f864322ab1a20cb00d7604/PhysicsTools/PatAlgos/src/MuonMvaEstimator.cc#L70 ). This deserves further studies in terms of muon selection optimality before changing it which goes slightly out of scope for this PR. Should we put it as a separate issue?

Reply:
> Should we put it as a separate issue?

OK.
Thank you for checking.